ingestion/src/legacy/raw.rs

```diff
@@ -0,0 +1,421 @@
+
+            let val = val.trim();
+
+            // // TODO: is this right?
+            // // Ignore empty values
+            if val.is_empty() {
+                continue;
+            }
+
+            let param = match param_entry {
+                Some(entry) => Param::Id(entry.id),
+                None => Param::Code(col.param_code),
+            };
+
+            let value: ObsType = if param_entry.is_some() && !param_entry.unwrap().is_scalar {
+                ObsType::NonScalar(val.to_string())
+            } else {
+                match val.parse::<f64>() {
+                    Ok(parsed) => ObsType::Scalar(parsed),
+                    Err(_) => ObsType::NonScalar(val.to_string()),
+                }
+            };
+
+            obs.push(UnlabelledDatum {
+                kvid: KvalobsId {
+                    station: header.station_id,
+                    param,
+                    typeid: header.type_id,
+                    sensor,
+                    level,
+                },
+                obstime: timestamp,
+                value,
+            })
+        }
+    }
+
+    Ok(obs)
+}
+
+// modified version of kldata::parse_kldata that returns RawDatum instead of ObsinnChunk
+pub fn parse(
+    msg: &str,
+    reference_params: ParamConversions,
+) -> Result<Vec<UnlabelledDatum>, ParseError> {
+    let (header, columns, csv_body) = {
+        let mut csv_body = msg.lines();
+
+        // parse the first two lines of the message as meta header, and csv column names,
+        // leave the rest as an iter over the lines of csv body
+        let header = ObsinnHeader::parse(csv_body.next().ok_or(ParseError::Lines)?)?;
+        let columns = parse_columns(csv_body.next().ok_or(ParseError::Lines)?)?;
+
+        (header, columns, csv_body)
+    };
+
+    parse_obs(csv_body, &columns, reference_params, header)
+}
+
+async fn insert(
+    conn: &mut PooledPgConn<'_>,
+    data: Vec<Datum>,
+    query: &Statement,
+    nonscalar_query: &Statement,
+) -> Result<(), Error> {
+    let transaction = conn.transaction().await?;
+
+    // This lock (and the one in the equivalent function in checked) are needed
+    // to prevent deadlocks between transactions from each db task.
+    //
+    // I think this (rare) deadlock happens because two transactions can try
+    // to update the same two rows in a different order, although one part of
+    // this that doesn't make sense is that the raw query doesn't update, and so
+    // shouldn't acquire a row lock? The docs are unclear on this.
+    //
+    // Unfortunately, the introduction of this lock seems to cause a ~30%
+    // slowdown If we want to reclaim that throughput, alternative approaches
+    // might be:
+    // - Just let the deadlocks happen, then catch them and retry
+    // - Order the queries in each transaction in a consistent way (sort by
+    //   timeseries then obstime?) so such a deadlock can't happen
```

Comment:
Would separating data in different vectors per obstype and using `COPY` have any benefit here?

Reply:
COPY is not option because it cannot handle conflicts

Reply:
Right, we would need to know for sure that the raw queue inserts first